pkg/logs/input/kubernetes/launcher.go

```diff
@@ -0,0 +1,220 @@
+		sources:            sources,
+		sourcesByContainer: make(map[string]*config.LogSource),
+		stopped:            make(chan struct{}),
+		kubeutil:           kubeutil,
+	}
+	err = launcher.setup()
+	if err != nil {
+		return nil, err
+	}
+	// Sources and services are added after the setup to avoid creating
+	// a channel that will lock the scheduler in case of setup failure
+	// FIXME(achntrl): Find a better way of choosing the right launcher
+	// between Docker and Kubernetes
+	launcher.dockerAddedServices = services.GetAddedServices(service.Docker)
+	launcher.dockerRemovedServices = services.GetRemovedServices(service.Docker)
+	launcher.containerdAddedServices = services.GetAddedServices(service.Containerd)
+	launcher.containerdRemovedServices = services.GetRemovedServices(service.Containerd)
+	return launcher, nil
+}
+
+// setup initializes the pod watcher and the tagger.
+func (l *Launcher) setup() error {
+	var err error
+	// initialize the tagger to collect container tags
+	err = tagger.Init()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Start starts the launcher
+func (l *Launcher) Start() {
+	log.Info("Starting Kubernetes launcher")
+	go l.run()
+}
+
+// Stop stops the launcher
+func (l *Launcher) Stop() {
+	log.Info("Stopping Kubernetes launcher")
+	l.stopped <- struct{}{}
+}
+
+// run handles new and deleted pods,
+// the kubernetes launcher consumes new and deleted services pushed by the autodiscovery
+func (l *Launcher) run() {
+	for {
+		select {
+		case service := <-l.dockerAddedServices:
+			l.addSource(service)
+		case service := <-l.dockerRemovedServices:
+			l.removeSource(service)
+		case service := <-l.containerdAddedServices:
+			l.addSource(service)
+		case service := <-l.containerdRemovedServices:
+			l.removeSource(service)
+		case <-l.stopped:
+			log.Info("Kubernetes launcher stopped")
+			return
+		}
+	}
+}
+
+// addSource creates a new log-source from a service by resolving the
+// pod linked to the entityID of the service
+func (l *Launcher) addSource(service *service.Service) {
+	pod, err := l.kubeutil.GetPodForEntityID(service.GetEntityID())
+	if err != nil {
+		log.Warnf("Could not add source for container %v: %v", service.Identifier, err)
+		return
+	}
+	container, err := l.searchContainer(service, pod)
+	if err != nil {
+		log.Debug(err)
+	}
+	source, err := l.getSource(pod, container)
+	if err != nil {
+		log.Warnf("Invalid configuration for pod %v, container %v: %v", pod.Metadata.Name, container.Name, err)
+		return
+	}
+	l.sourcesByContainer[container.ID] = source
+	l.sources.AddSource(source)
+}
+
+func (l *Launcher) searchContainer(service *service.Service, pod *kubelet.Pod) (kubelet.ContainerStatus, error) {
+	for _, container := range pod.Status.Containers {
+		containerID := container.ID
+		if service.GetEntityID() == containerID {
+			// If the container is already tailed, we don't do anything
+			// That shoudn't happen
+			if _, exists := l.sourcesByContainer[containerID]; exists {
```

Comment:
I don't think it should be the responsibility of the this method to assert that the source has already been processed. I would move this upstream, i.e in `addSource`.